etcd-client/src/watch/mod.rs

```diff
@@ -40,163 +40,545 @@
+        Self {
+            watched_map: Lazy::new(WatchedMap::default),
+            waiting_cancels: Lazy::new(HashMap::new),
+            queued_watch_requests: SegQueue::new(),
+            waiting_watch_request: Lazy::new(|| None),
+            cancel_req_sender,
+            shutdown,
+            sub_tasks: Some(SegQueue::new()),
+        }
+    }
+
+    /// cancel a watch in async task.
+    async fn cancel_watch(&self, keyrange: KeyRange) {
+        log::debug!("cancel watch {}", keyrange);
+        let mut watched_map = self.watched_map.write().await;
+        if let Some(watchid) = watched_map.remove_watch(&keyrange) {
+            let (tx, rx) = bounded::<()>(1);
+            self.waiting_cancels.write().await.insert(watchid, tx);
+            if self.cancel_req_sender.send(watchid).await.is_ok() {
+                self.sub_tasks
+                    .as_ref()
+                    .unwrap_or_else(|| { panic!("sub_tasks should be some until dropped") })
+                    .push(
+                        smol::spawn(async move {
+                            futures::select! {
+                        _ = smol::Timer::after(Duration::from_secs(WATCH_REQUEST_TIMEOUT_SEC)).into_future().fuse()=>{
+                            // todo: add retry for failed request
+                            // return Err(EtcdError::WaitingResponseTimeout("waiting for cancel response when calling `cancel_watch`".to_owned()));
+                            log::debug!("cancel watch wait response timeout");
+                        }
+                        res = rx.recv().into_future().fuse()=>{
+                            res.unwrap_or_else(|e|{
+                                panic!("receive cancel response channel shouldn't be destroyed, err:{e}");
+                            });
+                            log::debug!("cancel watch successed");
+                        }
+                    }
+                        })
+                    );
+            }
+        } else {
+            panic!("logic bug, cancel watch should be called only when there's watched key");
+        }
+    }
+}
 /// `WatchTunnel` is a reusable connection for `Watch` operation
 /// The underlying `gRPC` method is Bi-directional streaming
+#[allow(dead_code)]
 struct WatchTunnel {
-    /// A channel sender to send watch request.
-    req_sender: Sender<EtcdWatchRequest>,
+    /// A channel sender to send watch request to send loop.
+    watch_req_sender: Sender<(EtcdWatchRequest, Sender<Option<SingleWatchEventReceiver>>)>,
     /// A channel receiver to receive watch response.
-    resp_receiver: Option<Receiver<Result<EtcdWatchResponse>>>,
-    /// A channel sender to send shutdowm request.
-    shutdown: Option<Sender<()>>,
+    // resp_receiver: Option<Receiver<Result<EtcdWatchResponse>>>,
+
+    /// Shared
+    shared: Arc<WatchTunnelShared>,
 }
 
 impl WatchTunnel {
+    #[allow(clippy::too_many_lines)]
```

Comment:
Try to break it into multiple functions.

Reply:
Ok.